A54-restrict-control-plane-status-codes.md

```diff
@@ -0,0 +1,136 @@
+
+## Abstract
+
+The gRPC control plane has no need to use many of the status codes yet
+communicates with external services whose status codes may accidentally be
+copied to the data plane. This gRFC restricts the allowed status codes the
+control plane can use when failing an RPC, as a safeguard against bugs.
+
+## Background
+
+gRPC has a control plane consisting mainly of name resolvers and load balancers.
+Separately it has a data plane that processes individual RPCs. Error information
+from the control plane needs to pass to the data plane when errors are
+occurring, but the gRPC data plane is restricted from generating [status codes
+reserved for application use][statuscodes.md]: `INVALID_ARGUMENT`, `NOT_FOUND`,
+`ALREADY_EXISTS`, `FAILED_PRECONDITION`, `ABORTED`, `OUT_OF_RANGE`, and
+`DATA_LOSS`. Those status codes are reserved for each service.
+
+Earlier in gRPC's life, there was relatively low risk of using a reserved status
+code in the data plane. There was a relatively fixed amount of control plane
+code, and much of it wasn't accessing gRPC-based services (e.g., DNS,
+pick_first, round_robin). At that point in time, a simple audit could reveal
+broken status code usages.
+
+Now, however, it is more common to implement custom name resolvers and load
+balancers and for those components to use gRPC themselves. This happened first
+with gRPC-LB, but also happens with xDS, Route Lookup Service (RLS), and others.
+It is now harder to audit all the code running in the control plane and it is
+very hard to find status "leaks" from a remote control plane server into the
+data plane. This is a serious risk as control plane errors can't be copied
+verbatim to the data plane; they need to be reinterpreted for the new context
+(generally by becoming `UNAVAILABLE` and attaching additional context).
+
+As an example, if a name resolver is looking up the host `example.com`, that
+could be implemented as a gRPC call to a service. That service might fail the
+RPC with `NOT_FOUND` because the host is not registered. But if the data plane
+has an RPC reading a file `funny-jokes.txt`, the RPC should fail with
+`UNAVAILABLE`. Failing the RPC with `NOT_FOUND` would confuse the application
+into believing `funny-jokes.txt` doesn't exist. This is the situation that
+caused a serious [Spotify outage][] because of [a bug in gRPC Java][issue 8950].
+
+[statuscodes.md]: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+[Spotify outage]: https://engineering.atspotify.com/2022/03/incident-report-spotify-outage-on-march-8/
+[issue 8950]: https://github.com/grpc/grpc-java/issues/8950
+
+## Proposal
+
+Add checks to each language's Channel that prevents the control plane from
+passing "inappropriate statuses" to the data plane. Specifically:
+
+ * When applying a failed Service Config parsing result from the name resolver
+ * When applying a failed Config Selector result from the name resolver (an
```

Comment:
This can reference [gRFC A31](https://github.com/grpc/proposal/blob/master/A31-xds-timeout-support-and-config-selector.md), where the config selector was specified.